src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,304 @@
+
+    const packageJson: PackageJsonData = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package.json'),
+    );
+
+    const packageLockJson: any = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package-lock.json'),
+    );
+
+    // They don't have a package.json for some reason? Eject button.
+    if (!!packageJson) {
+      return [];
+    }
+
+    const packages = Object.entries({
+      ...packageJson.devDependencies,
+      ...packageJson.dependencies,
+    }) as [string, string][];
+
+    // Collect packages only in the stencil, ionic, or capacitor org's:
+    // https://www.npmjs.com/org/stencil
+    const ionicPackages = packages.filter(
+      ([k]) => k.startsWith('@stencil/') || k.startsWith('@ionic/') || k.startsWith('@capacitor/'),
+    );
+
+    const versions = packageLockJson
+      ? ionicPackages.map(
+          ([k, v]) =>
+            `${k}@${
+              (packageLockJson?.dependencies[k] as { version: string })?.version ??
+              (packageLockJson?.devDependencies[k] as { version: string })?.version ??
+              v
+            }`,
+        )
+      : ionicPackages.map(([k, v]) => `${k}@${v}`);
+
+    return versions;
+  } catch (err) {
+    hasDebug() && console.error(err);
+    return [];
+  }
+}
+
+/**
+ * If telemetry is enabled, send a metric via IPC to a forked process for uploading.
+ */
+export async function sendMetric(name: string, value: TrackableData): Promise<void> {
+  const message: Metric = {
+    name,
+    timestamp: new Date().toISOString(),
+    source: 'stencil_cli',
+    value,
+  };
+
+  await sendTelemetry({ type: 'telemetry', message });
+}
+
+/**
+ * Used to read the ~/.ionic/config.json file's tokens.telemetry property.
+ * @returns string
+ */
+async function getTelemetryToken() {
+  const config = await readConfig();
+  return config['tokens.telemetry'];
+}
+
+/**
+ * Issues a request to the telemetry server.
+ * @param data Data to be tracked
+ */
+async function sendTelemetry(data: any) {
+  try {
+    const now = new Date().toISOString();
+
+    const body = {
+      metrics: [data.message],
+      sent_at: now,
+    };
+
+    // This request is only made if telemetry is on.
+    const response = await getCompilerSystem().fetch('https://api.ionicjs.com/events/metrics', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(body),
+    });
+
+    hasVerbose() &&
+      console.debug('\nSent %O metric to events service (status: %O)', data.message.name, response.status, '\n');
+
+    if (response.status !== 204) {
+      hasVerbose() &&
+        console.debug('\nBad response from events service. Request body: %O', response.body.toString(), '\n');
+    }
+  } catch (e) {
+    hasVerbose() && console.debug('Telemetry request failed:', e);
+  }
+}
+
+/**
+ * Checks if telemetry is enabled
+ * @returns boolean
```

Comment:
See my comment above regarding having just the type in `@returns`. I think we should strive to be consistent here - we should decide _what_ we want to do for `@returns` and apply that uniformly here and to the `enableTelemetry` and `disableTelemetry` methods below

Reply:
Adde